### ISSUE TITLE
Audio latency and UI issues on Apple devices

### ISSUE THREAD

**Issue by vundurty** [Scirra/Construct-3-bugs#4793]
## Problem description

We are not sure whether the issues that we are experiencing are bugs or not. Therefore, we are reaching out to you to ask for a confirmation.

We have noticed two issues

1. Audio issue - on iOS app, the audio plays without any time delay but on Safari browser for MacOS, a stereo track incorrectly plays in mono and is delayed before starting. We have tested it as an Android game, Chrome browser for Windows and Chrome browser for MacOS, it worked perfectly fine.

2. Slider issue - we have setup a slider (it's neither the Construct 3 plugin nor any third party plugin, it was setup manually) in the game for providing multiple options for the user to choose from. The slider bar range was offset or shifted to the left on the iOS app (we were able to test the game on iPad Pro (2nd Generation), iPhone SE (1st Generation), and iPhone SE (2nd Generation)) whereas it is adjusted perfectly fine on Safari browser for MacOS, Chrome browser for MacOS, Chrome browser for Windows and Android app.

We are not sure what might be the issue. We have contacted the Apple engineers first before posting here, they have suggested us to contact Scirra.

Please let us know whether the observations and conclusions of issues provided are bugs on your end or not. For instance, if they are programmable then, please let us know what to adjust, or if they are Apple bugs, then we need to contact Apple engineers.

I have attached the C3P file for your review.

## Attach a .c3p

[Apple C3 Test.zip](https://github.com/Scirra/Construct-3-bugs/files/6412523/Apple.C3.Test.zip)

## Steps to reproduce

For audio issue
1. Preview project on Safari for MacOS
2. You will notice the audio latency

For UI issue
1. Export the project to XCode
2. Preview the game on XCode
3. You will notice the slider bar range was offset or shifted to the left on the iOS app 

## Observed result

For audio issue - Audio doesn't play immediately, it take a little time
For UI issue - The slider bar range was offset or shifted to the left (X coordinates)

## Expected result

For audio issue - Audio to be played immediately without any latency
For UI issue - The slider should be in the middle of the screen (X coordinates)

## More details

**Affected browsers/platforms:** 
Safari, MacOS - Affected
iOS - we were able to test the game on iPad Pro (2nd Generation), iPhone SE (1st Generation), and iPhone SE (2nd Generation)
Chrome, MacOS - Works fine, no issues
Chrome, Windows - Works fine, no issues
Android - Works fine, no issues

**First affected release:** 
We have noticed this issue in almost all releases. We have noticed this more than an year back but we thought it could be a programming issue, so I'm not able to provide the exact release but as of now it persists in r241 stable

## System details

<details><summary>View details</summary>

Platform information
Browser: Chrome
Browser version: 90.0.4430.93
Browser engine: Chromium
Browser architecture: 64-bit
Context: browser
Operating system: Windows
Operating system version: 10
Operating system architecture: 64-bit
Device type: desktop
Device pixel ratio: 1
Logical CPU cores: 4
Approx. device memory: 8 GB
User agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.93 Safari/537.36
C3 release: r241 (stable)
Language setting: en-US

Local storage
Storage quota (approx): 108 gb
Storage usage (approx): 365 mb (0.3%)
Persistant storage: No

Browser support notes
This list contains missing features that are not required, but could improve performance or user experience if supported.

Nothing is missing. Everything is OK!
WebGL information
Version string: WebGL 2.0 (OpenGL ES 3.0 Chromium)
Numeric version: 2
Supports NPOT textures: yes
Supports GPU profiling: yes
Supports highp precision: yes
Vendor: Google Inc. (Intel)
Renderer: ANGLE (Intel, Intel(R) HD Graphics Family Direct3D11 vs_5_0 ps_5_0, D3D11-10.18.15.4278)
Major performance caveat: no
Maximum texture size: 16384
Point size range: 1 to 1024
Extensions:

EXT_color_buffer_float
EXT_color_buffer_half_float
EXT_disjoint_timer_query_webgl2
EXT_float_blend
EXT_texture_compression_bptc
EXT_texture_compression_rgtc
EXT_texture_filter_anisotropic
EXT_texture_norm16
KHR_parallel_shader_compile
OES_texture_float_linear
WEBGL_compressed_texture_s3tc
WEBGL_compressed_texture_s3tc_srgb
WEBGL_debug_renderer_info
WEBGL_debug_shaders
WEBGL_lose_context
WEBGL_multi_draw
OVR_multiview2
Audio information
System sample rate: 48000 Hz
Output channels: 2
Output interpretation: speakers
Supported decode formats:

WebM Opus (audio/webm; codecs=opus)
Ogg Opus (audio/ogg; codecs=opus)
WebM Vorbis (audio/webm; codecs=vorbis)
Ogg Vorbis (audio/ogg; codecs=vorbis)
MPEG-4 AAC (audio/mp4; codecs=mp4a.40.5)
MP3 (audio/mpeg)
FLAC (audio/flac)
PCM WAV (audio/wav; codecs=1)
Supported encode formats:

WebM Opus (audio/webm; codecs=opus)
Video information
Supported decode formats:

WebM AV1 (video/webm; codecs=av01.0.00M.08)
MP4 AV1 (video/mp4; codecs=av01.0.00M.08)
WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)
Ogg Theora (video/ogg; codecs=theora)
H.264 (video/mp4; codecs=avc1.42E01E)
Supported encode formats:

WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)

</details>

**Comment by AshleyScirra** [comment]
These kinds of issues are difficult to investigate - this combines two issues in one go, and we need a minimal project for each individual issue. It would be better to file separate issues for each issue and follow all the bug report guidelines.

For the Audio issue, I can't reproduce any latency issue in Safari on macOS, if I click a small circle it plays a sound immediately, although given all the stuff in the project it's not clear if this is what I'm meant to do. Also note browsers prevent audio playback until the user first interacts with the page in order to avoid web pages annoying the user, so you should assume the page will be muted until the first user interaction. WebM Opus audio will play in mono in Safari since Safari is the only browser that does not support WebM Opus, so we have to ship our own decoder, and at the moment that decoder only decodes in mono. This is one of the many complications that arises from Safari's lack of support for WebM Opus that all other browsers support. By far the best thing to do would be for Apple to add support for WebM Opus - you could also voice your support for it here: https://bugs.webkit.org/show_bug.cgi?id=183852

As for the slider issue it just seems too complicated to tell because your events seem to do a bunch of calculations in different circumstances affecting the viewport. It could be your events that are wrong. This is why we ask for a minimal project, to make sure it's not your events or project that is wrong. It also looks like the default viewport is larger than the layout size, which is an unusual setup and might look wrong - you should make sure you always have a layout at least the same size as the viewport, or enable unbounded scrolling so the layout scroll boundaries don't interfere unexpectedly with the scroll position. I'd strongly recommend just using one of the built-in fullscreen modes and not trying to set up complicated events that try to manage the aspect ratio and canvas size manually. It's a really complicated area and Construct does a lot behind-the-scenes to make sure the built in options "just work".

In short I can't see any good evidence that Construct is doing something wrong in either case, so closing. Please submit individual issues following all the guidelines if you still think something is wrong in either case.
